File: 6.zigzag-conversion/solution.cpp
class Solution {
public:
    string convert(string s, int numRows) {
		if (numRows == 1)
			return s;
		int len = s.length();
		string ans;
		for (int i = 0; i < numRows; i++)
		{
			if (i >= len)
				break;
			ans += s[i];
			for (int j = 1;; j++)
			{
				if (i != 0 && i != numRows - 1)
				{
					int index = j * (numRows - 1) * 2 - i;
					if (index >= len)
						break;
					ans += s[index];
				}
				int index = i + j * (numRows - 1) * 2;
				if (index >= len)
					break;
				ans += s[index];
			}
		}
		return ans;
    }
};

File: 6.zigzag-conversion/solution.kt
class Solution {
    fun convert(s: String, numRows: Int): String {
        if (numRows == 1) return s
        val m = Array(numRows) { CharArray(s.length) }
        var up = false
        var i = 0
        var j = 0
        for (c in s) {
            m[i][j] = c
            if (!up) {
                if (++i == numRows - 1)
                    up = true
            } else {
                j++
                if (--i == 0)
                    up = false
            }
        }
        val sb = StringBuilder()
        for (n in m) {
            for (k in n) {
                if (k != '\u0000') sb.append(k)
            }
        }
        return sb.toString()
    }
}
